File: dotnet/Tybscri/Symbols/SourceSymbol.cs
using System.Linq.Expressions;

namespace Tybscri.Symbols;

public class SourceSymbol : ISymbol
{
    private readonly ISymbolDefinitionNode _node;

    public SourceSymbol(string name, ISymbolDefinitionNode node)
    {
        Name = name;
        _node = node;
    }

    public TybscriType ValueType => _node.SymbolType;

    public string Name { get; }

    public Expression LinqExpression => _node.LinqExpression;

    public void Resolve()
    {
        _node.ResolveSymbolDefinition();
    }
}

File: js/lang/src/SourceSymbol.ts
import { ResolveContext } from "./common";
import { Symbol } from "./Symbol";
import { Type } from "./typeSystem";

export class SourceSymbol extends Symbol {
  public get isConst(): boolean {
    return this.node.isConst;
  }

  constructor(
    name: string,
    public readonly node: {
      resolve(context: ResolveContext): void;
      valueType: Type;
      isConst: boolean;
    }
  ) {
    super(name);
  }

  public get valueType(): Type {
    return this.node.valueType;
  }

  public resolve(context: ResolveContext): void {
    this.node.resolve(context);
  }
}
